hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -414,6 +416,21 @@ object InferPType {
 
         branchType.setRequired(branchType.required && cond.pType2.required)
       }
+
+      case Coalesce(values) => {
+        InferPType(values(0), env)
+        var pTypes = ArrayBuffer[PType](values(0)._pType2)
```

Comment:
I think this is clearer if written functionally:

```scala
 getNestedElementPTypes(ptypes = values.map { v => 
    InferPType(v, env)
    v.pType2
})
```

Reply:
What about the assertion that all are isOfType?

Reply:
I could write this with a branch in the map, e.g

```scala
var head: IR: _
var i = 0;
 getNestedElementPTypes(ptypes = values.map { v => 
    InferPType(v, env)
    if(i > 0) assert(v._pType2 isOfType head._pType2)
    v.pType2
    i += 1
})
```

But I think this is less clear, this mixture of imperative and functional code.

Reply:
getNestedElementPTypes should do that assertion, and it should probably be an assertion that all the virtual types are the same.